M365/MDO/ResendFailedMail.ps1

```diff
@@ -0,0 +1,330 @@
+                $recipientCache[$failedMessage.SenderAddress] = $failedMessage.SenderAddress
+                Write-Verbose "Verified $($failedMessage.SenderAddress)"
+                $verifiedAcceptedSenderMessages += $failedMessage
+            }
+        } catch {
+            Write-Host "Error getting Sender $($failedMessage.SenderAddress)" -ForegroundColor Yellow
+            Write-Host "Discarded $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)" -ForegroundColor Yellow
+            Write-Verbose "$_"
+        }
+    }
+}
+
+if ( $verifiedAcceptedSenderMessages ) {
+    if ($verifiedAcceptedSenderMessages.Count) {
+        $totalMessages = $verifiedAcceptedSenderMessages.Count
+    } else {
+        $totalMessages = 1
+    }
+
+    if (-not $Force) {
+        Write-Host "`nWe are going to resend the following messages:"
+        Write-Host ($verifiedAcceptedSenderMessages | Format-Table -AutoSize Received, MessageId, SenderAddress, RecipientAddress, Subject | Out-String)
+        Write-Host "Total number of messages: $totalMessages`n"
+    }
+
+    if ($Force -or $PSCmdlet.ShouldContinue("Are you sure you want to do it?", "Resend messages")) {
+        $count = 0
+        foreach ( $failedMessage in $verifiedAcceptedSenderMessages ) {
+            $count++
+            Write-Progress -Activity "Resending Progress" -Status "$count of $totalMessages" -PercentComplete ($count / $totalMessages * 100) -CurrentOperation "Resending message $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+            try {
+                $fullMessage = $null
+                $fullMessage = Get-MgUserMessage -UserId $failedMessage.SenderAddress -Filter "InternetMessageId eq '$($failedMessage.MessageId)'" -ExpandProperty Attachments -ErrorAction Stop | Sort-Object ReceivedDateTime | Select-Object -First 1
+            } catch {
+                Write-Host "Error getting message $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)" -ForegroundColor Red
+                Write-Verbose "$_"
+            }
+            if ($fullMessage.Count -eq 0) {
+                Write-Host "Message not found for $($failedMessage.MessageId)" -ForegroundColor Yellow
+            } else {
+                Write-Verbose "Resending message $($failedMessage.MessageId) - Subject: $($fullMessage.Subject)"
+                try {
+                    #Send-MgUserMail -UserId $failedMessage.SenderAddress -Message $fullMessage -ErrorAction Stop
+                    Send-MgUserMessage -UserId $failedMessage.SenderAddress -MessageId $fullMessage.Id -ErrorAction Stop
+                    Write-Host "Resent Message: $($failedMessage.MessageId) - Subject: $($fullMessage.Subject)"
+                } catch {
+                    Write-Host "Error resending message $($failedMessage.MessageId) - Subject: $($fullMessage.Subject)" -ForegroundColor Red
+                    Write-Verbose "$_"
+                }
+            }
+        }
+    }
```

Comment:
At the end, we should provide the count of the successful and failed to resends.

Reply:
Ok, will do.